src/chainstate/burn/operations/peg_in.rs

```diff
@@ -0,0 +1,470 @@
+        }
+
+        let (amount, peg_wallet_address) =
+            if let Some(Some(recipient)) = tx.get_recipients().first() {
+                (recipient.amount, recipient.address.clone())
+            } else {
+                warn!("Invalid tx: Output 2 not provided");
+                return Err(OpError::InvalidInput);
+            };
+
+        let parsed_data = Self::parse_data(&tx.data())?;
+
+        let txid = tx.txid();
+        let vtxindex = tx.vtxindex();
+        let block_height = block_header.block_height;
+        let burn_header_hash = block_header.block_hash;
+
+        Ok(Self {
+            recipient: parsed_data.recipient,
+            peg_wallet_address,
+            amount,
+            memo: parsed_data.memo,
+            txid,
+            vtxindex,
+            block_height,
+            burn_header_hash,
+        })
+    }
+
+    fn parse_data(data: &[u8]) -> Result<ParsedData, ParseError> {
+        /*
+            Wire format:
+
+            0      2  3                  24                            64       80
+            |------|--|------------------|-----------------------------|--------|
+             magic  op   Stacks address      Contract name (optional)     memo
+
+             Note that `data` is missing the first 3 bytes -- the magic and op must
+             be stripped before this method is called. At the time of writing,
+             this is done in `burnchains::bitcoin::blocks::BitcoinBlockParser::parse_data`.
+        */
+
+        if data.len() < 21 {
+            warn!(
+                "PegInOp payload is malformed ({} bytes, expected at least {})",
+                data.len(),
+                21
+            );
+            return Err(ParseError::MalformedData);
+        }
+
+        let version = *data.get(0).ok_or(ParseError::MalformedData)?;
+        let address_data: [u8; 20] = data
+            .get(1..21)
+            .ok_or(ParseError::MalformedData)?
+            .try_into()?;
+
+        let standard_principal_data = StandardPrincipalData(version, address_data);
+
+        let memo = data.get(61..).unwrap_or(&[]).to_vec();
+
+        let recipient: PrincipalData =
+            if let Some(contract_bytes) = Self::leading_non_zero_bytes(data, 21, 61) {
+                let contract_name: String = std::str::from_utf8(contract_bytes)?.to_owned();
+
+                QualifiedContractIdentifier::new(standard_principal_data, contract_name.try_into()?)
+                    .into()
+            } else {
+                standard_principal_data.into()
+            };
+
+        Ok(ParsedData { recipient, memo })
+    }
+
+    pub fn check(&self) -> Result<(), OpError> {
+        if self.amount == 0 {
+            warn!("PEG_IN Invalid: Peg amount must be positive");
+            return Err(OpError::PegInAmountMustBePositive);
+        }
+
+        Ok(())
+    }
+
+    fn leading_non_zero_bytes(data: &[u8], from: usize, to: usize) -> Option<&[u8]> {
```

Comment:
A defensive check that `from < to` might be appropriate here.

Reply:
Makes sense. I added an assert since an error would indicate a severe bug in the code at which point a panic seems warranted. Let me know if you prefer a normal error here instead.